airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/destination_async/buffers/BufferDequeue.java

```diff
@@ -0,0 +1,127 @@
+ * queue metadata required to determine buffer flushing.
+ */
+// todo (cgardens) - make all the metadata methods more efficient.
+public class BufferDequeue {
+
+  private final GlobalMemoryManager memoryManager;
+  private final ConcurrentMap<StreamDescriptor, MemoryBoundedLinkedBlockingQueue<AirbyteMessage>> buffers;
+  private final ConcurrentMap<StreamDescriptor, ReentrantLock> bufferLocks;
+
+  public BufferDequeue(final GlobalMemoryManager memoryManager,
+                       final ConcurrentMap<StreamDescriptor, MemoryBoundedLinkedBlockingQueue<AirbyteMessage>> buffers) {
+    this.memoryManager = memoryManager;
+    this.buffers = buffers;
+    bufferLocks = new ConcurrentHashMap<>();
+  }
+
+  /**
+   * Primary dequeue method. Best-effort read a specified optimal memory size from the queue.
+   *
+   * @param streamDescriptor specific buffer to take from
+   * @param optimalBytesToRead bytes to read, if possible
+   * @return
+   */
+  public Batch take(final StreamDescriptor streamDescriptor, final long optimalBytesToRead) {
+    final var queue = buffers.get(streamDescriptor);
+
+    if (!bufferLocks.containsKey(streamDescriptor)) {
+      bufferLocks.put(streamDescriptor, new ReentrantLock());
+    }
+
+    bufferLocks.get(streamDescriptor).lock();
+    try {
+      final AtomicLong bytesRead = new AtomicLong();
+
+      final var s = Stream.generate(() -> {
+        try {
+          return queue.poll(5, TimeUnit.MILLISECONDS);
+        } catch (final InterruptedException e) {
+          throw new RuntimeException(e);
+        }
+      }).takeWhile(memoryItem -> {
+        // if no new records after waiting, the stream is done.
+        if (memoryItem == null) {
+          return false;
+        }
+
+        // otherwise pull records until we hit the memory limit.
+        final long newSize = memoryItem.size() + bytesRead.get();
+        if (newSize <= optimalBytesToRead) {
+          bytesRead.addAndGet(memoryItem.size());
+          return true;
+        } else {
+          return false;
+        }
+      }).map(MemoryBoundedLinkedBlockingQueue.MemoryItem::item)
+          .toList()
+          .stream();
+
+      // todo (cgardens) - possible race where in between pulling records and new records going in that we
+      // reset the limit to be lower than number of bytes already in the queue. probably not a big deal.
+      queue.setMaxMemoryUsage(queue.getMaxMemoryUsage() - bytesRead.get());
+
+      return new Batch(s, bytesRead.get(), memoryManager);
+    } finally {
+      bufferLocks.get(streamDescriptor).unlock();
+    }
+  }
+
+  /**
+   * The following methods are provide metadata for buffer flushing calculations. Consumers are
+   * expected to call {@link #getBufferedStreams()} to retrieve the currently buffered streams as a
+   * handle to the remaining methods.
+   */
+
+  public Set<StreamDescriptor> getBufferedStreams() {
```

Comment:
this returns a set instead of the underlying map, letting the memory bounded queue by package private.